R/permute_cases.R

```diff
@@ -21,38 +21,56 @@ permute_cases.data.frame <- function(cases, n_permutations, feature_distribution
   perm[seq.int(1, by = n_permutations, length.out = nrow(cases)), ] <- cases
   perm
 }
-#' @importFrom stringi stri_split stri_join_list
+
 #' @importFrom Matrix Matrix
-permute_cases.character <- function(cases, n_permutations, split_by, bow) {
-  tokens <- stri_split(cases, regex = split_by)
-  doc_size <- lengths(tokens)
-  if (bow) {
-    tokens_lower <- split(tolower(unlist(tokens)), rep(seq_along(tokens), doc_size))
-    tokens_lookup <- lapply(tokens_lower, unique)
-    doc_size <- lengths(lapply(tokens_lookup, unique))
-  } else {
-    tokens_lookup <- tokens
-  }
-  perm <- lapply(seq_along(tokens_lookup), function(i) {
-    remove_n <- sample(doc_size[i], n_permutations, replace = FALSE)
-    remove <- lapply(remove_n, sample, x = doc_size[i] - 1)
-    permutations <- lapply(remove, function(ii) tokens_lookup[[i]][-ii])
-    if (bow) {
-      permutations <- lapply(permutations, function(words) tokens[[i]][!tokens_lower[[i]] %in% words])
-      var_names <- tokens_lookup[[i]]
-    } else {
-      var_names <- paste0(seq_along(tokens_lookup[[i]]), '_', tokens_lookup[[i]])
-    }
-    complete <- rep(1, length(tokens_lookup[[i]]))
-    remove <- do.call(rbind, lapply(remove, function(ii) {complete[ii] <- 0; complete}))
-    remove[1, ] <- 1
-    remove <- Matrix(remove, sparse = TRUE)
-    colnames(remove) <- tokens_lookup[[i]]
-    list(permutations, remove)
-  })
-  removes <- lapply(perm, `[[`, 2)
-  perm <- lapply(perm, `[[`, 1)
-  perm <- stri_join_list(unlist(perm, recursive = FALSE), sep = ' ')
-  perm[seq.int(1, by = n_permutations, length.out = length(cases))] <- cases
-  list(permutations = perm, tabular = removes)
+#' @importFrom purrr map map2 flatten_chr set_names flatten flatten_int map_chr flatten_dbl
+#' @importFrom stringdist seq_dist
+#' @importFrom magrittr %>% set_colnames %<>%
+permute_cases.character <- function(cases, n_permutations, tokenization, bow, dist_fun) {
+  documents_tokens <- map(cases, tokenization) %>%
+{d_tokens <- . ;
+  map2(d_tokens, lengths(d_tokens) %>% cumsum() %>% head(., length(.) - 1) %>% c(0, .),
+       ~ {if (bow) paste0(.x, "_",  seq_along(.) + .y) else unique(.x)})}
+
+  tokens <- documents_tokens %>%
+    flatten_chr() %>%
+    unique(.) %>%
+    sort(decreasing = FALSE) %>%
+    set_names(seq(.))
+
+  documents_tokens %<>%
+    map(~ {d <- . ; which(tokens %in% d) %>%
+      set_names(d, nm = .)})
+
+  dict_size <- length(tokens)
+
```

Comment:
Personal preference is to avoid `%<>%` inside packages as it is a bit less clear whats going on (it is not a clear assignment and R itself doesn't understand it as such). Can we keep it to using the `%>%` pipe?